### PR TITLE
fix for AppCLoud-147 for version creation

### PR DESCRIPTION
**Pull request by SabraO**

### CODE REVIEW

File: modules/jaggeryapps/appmgt/src/site/blocks/application/application.jag [hunk 1]
```diff
@@ -76,9 +76,19 @@ var log = new Log();
             response.status = 400; // bad request
             response.message = "Application with same name and version already exists!";
         } else {
-            var applicationCount = applicationModule.getApplicationCount();
-            var maxAppCount = applicationModule.getMaxInstanceCount();
-            if (applicationCount < maxAppCount) {
+            var isAppCountLowerThanMax = false;
+
+            if (isNewVersion) {
+                isAppCountLowerThanMax = true;
```
Comment (manjulaRathnayaka) [inline modules/jaggeryapps/appmgt/src/site/blocks/application/application.jag]: Add a description here explaining the reason.

Comment (SabraO) [inline modules/jaggeryapps/appmgt/src/site/blocks/application/application.jag]: If it is a new version of an existing app then the application count is not checked to see if is lower than the maximum. If it is a new application (hence not a new version) then the application count check is conducted. 
If it is a new version but the application does not exist then app creation fails.

File: modules/jaggeryapps/appmgt/src/site/blocks/application/application.jag [hunk 2]
```diff
@@ -76,9 +76,33 @@ var log = new Log();
...
+                if (applicationModule.getApplicationHashIdByName(applicationName) != null) {
+                    isAppCountLowerThanMax = true;
+                } else {
+                    log.warn("Application creation request not accepted because" + msg + " no application with " +
+                            applicationName + " exists.");
+                }
+            } else {
+                //Since the user is trying to create a new application we check if the existing application count is less
+                //than the maximum amount of applications that can be created
+                var applicationCount = applicationModule.getApplicationCount();
```
Comment (danushkaf) [inline modules/jaggeryapps/appmgt/src/site/blocks/application/application.jag]: Lets fix the getApplicationCount also. Currently its returning no of versions. Lets fix it to count no of applications.

Comment (SabraO) [inline modules/jaggeryapps/appmgt/src/site/blocks/application/application.jag]: getApplicationCount returns the number of applications. It was modified earlier for the requirement of displaying the banner.

File: modules/jaggeryapps/appmgt/src/site/blocks/application/application.jag (hunk 1 above)
Comment (SabraO) [inline modules/jaggeryapps/appmgt/src/site/blocks/application/application.jag]: fixed at cefd7b4f0fbcee507574e856c14c77c02605558a

File: modules/jaggeryapps/appmgt/src/site/blocks/application/application.jag [hunk 3]
```diff
@@ -76,9 +76,38 @@ var log = new Log();
...
         } else {
-            var applicationCount = applicationModule.getApplicationCount();
-            var maxAppCount = applicationModule.getMaxInstanceCount();
-            if (applicationCount < maxAppCount) {
+            var isAppCountLowerThanMax = false;
+
+            //Added to restrict application creation using the REST api
+            //If the user is trying to create a new version first check if an application with the given name exists
+            //before creating the new application version
+            if (isNewVersion) {
```
Comment (chalitha1989) [inline modules/jaggeryapps/appmgt/src/site/blocks/application/application.jag]: simplify the 'isNewVersion' checking logics and do it in the top.

**Pull request merged**